src/OpenTelemetry.Api/Internal/Shims/RequiresDynamicCodeAttribute.cs

```diff
@@ -0,0 +1,55 @@
+// <copyright file="RequiresDynamicCodeAttribute.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+#if !NET7_0_OR_GREATER
```

Comment:
Don't we need a `net7.0` target for this to work?

Reply:
This attribute is in the runtime of .net7.0 and .net8.0: https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.requiresdynamiccodeattribute?view=net-7.0

Therefore, we will only need to compile this file if it is not net7.0 or greater.

Reply:
It's basically future proofing the code. That way if/when we add a net7.0+ target, this code doesn't get compiled for it. Yes it isn't strictly necessary now, but it is saving some effort/confusion for a future dev who adds a new TFM.

Reply:
This kind of forces us to always add the new target to the API project as well. If in the future, the SDK project also ends up using this attribute, we wouldn't be able to add a net7.0+ target **only** to the SDK project as it would complain about the attribute existing in both `OpenTelemetry.Api` (because of InternalsVisible) and `System.Runtime`.

Reply:
Do you have a recommendation to resolve this? The only other option I could imagine is to `#if NET7_OR_GREATER` where we need to use this attribute AND add a `net7.0` target now in order to use the attribute.

> This kind of forces us to always add the new target to the API project as well

Using InternalsVisibleTo across separate NuGet packages isn't recommended, since the package versions can get out of sync and users will see errors.

In general, matching TFMs across these packages is preferrable. It makes maintenance much easier. It is how we manage our packages we ship out of dotnet/runtime. They all have a consistent set of targets. For the 8.0-* packages it is: `net6.0;net7.0;net8.0;netstandard2.0;net462`. For example see https://www.nuget.org/packages/System.Collections.Immutable/8.0.0-preview.4.23259.5#dependencies-body-tab:

![image](https://github.com/open-telemetry/opentelemetry-dotnet/assets/8291187/b6b2f80e-bd2a-446b-ba10-b3c643deecf0)